### PR TITLE
[NEXUS-8014][CMA] Repository remote connection status/availability

### PR DESCRIPTION
**Pull request by adreghiciu**
https://issues.sonatype.org/browse/NEXUS-8014

Remote availability is checked at http client level. This ensures that availability check is done / status is updated regardless from where remote is accessed(e.g.ProxyFacetSupport, NugetFeedFetcher), and without interfering with the code that makes the access.

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/RemoteConnectionStatus.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.httpclient;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @since 3.0
+ */
+public class RemoteConnectionStatus
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/RemoteConnectionStatus.java]: Don't we need some `isAvailable` boolean method on this? As currently only description/reason differentiates statuses, that I guess are mostly for UI...

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+          synchronized (this) {
+            // avoid some other thread already increased the sequence
+            if (blockedUntil == null || blockedUntil.isBeforeNow()) {
+              blockedUntil = DateTime.now().plus(autoBlockSequence.next());
+              if (checkThread != null) {
+                checkThread.interrupt();
+              }
+              String uri = target.toURI();
+              // TODO maybe find different means to schedule status checking
+              checkThread = new Thread(new CheckStatus(uri, blockedUntil), "Check Status " + uri);
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java]: In NX2 repositories had a thread pool for tasks like these. This not only helped centralising the work, but also helps a lot wrt thread naming (to be repo specific) and makes better logging, easier to connect status checking thread with repo being checked.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    blocked = checkNotNull(config).getConnectionConfig().isBlocked();
+    autoBlock = config.getConnectionConfig().shouldAutoBlock();
+    status = new RemoteConnectionStatus(blocked ? "Remote Manually Blocked" : "Unknown");
+    // TODO shall we use config.getConnectionConfig().getTimeout() * 2 as in NX2?
+    autoBlockSequence = new FibonacciNumberSequence(Time.seconds(40).toMillis());
+  }
+
+  private <T> T filter(final HttpHost target, final Filterable<T> filterable) throws IOException {
+    // main target is the first accessed target
+    if (mainTarget == null) {
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java]: Concurrent calls might end up with wrong mainTarget set?

Comment (adreghiciu) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java]: I did not want to sync on this as in my view different threads for same repo they all start with same target so regardless the task they should all end up with same main target

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java (hunk 2 above)
Comment (adreghiciu) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java]: Discussed about this with Jason and we postponed having a thread pool till we have a better infrastructure in place. Put the TODO there so we revisit

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/RemoteConnectionStatus.java (hunk 1 above)
Comment (adreghiciu) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/RemoteConnectionStatus.java]: There is no use case for it yet so I did not add it

**Comment by kellyrob99** [comment]
+1

**Comment by cstamas** [comment]
+1

**Comment by adreghiciu** [comment]
Manually merged into nexus-cma3 https://github.com/sonatype/nexus-oss/commit/3d31fc0a30bcb967085afcf9bc0310143fb99f42

**Pull request closed without merging**